2018-10/workshop-two/workshop-two.md

```diff
@@ -0,0 +1,386 @@
+
+---
+
+# Workshop II Outline
+
+Create a web API for our previously created library
+
+* Using flags to configure app
+* http.Handler and http.HandlerFunc
+* Interfaces
+* JSON encoding 
+* Refactoring
+* Add tests for our API
+
+---
+
+# Development Environment
+
+Can use libraries such as Gin, to allow for livereloading
+
+1. https://github.com/codegangsta/gin
+
+```bash
+go get github.com/codegangsta/gin
+```
+
+---
+
+# Hello World
+
+* https://golang.org/dl/
+* https://golang.org/doc/install
+
+```go
+package main
+
+import "fmt"
+
+func main() {
+    fmt.Println("Hello, World")
+}
+```
+
+---
+
+# Flags
+
+* Go accepted way of configuration
+* Allows for default values
+* Provides documentation &rarr; `./myapp -help`
+
+```go
+import "flag"
+
+func main() {
+    var someVal string
+    flag.StringVar(&someVal, "someval", "", "some config value")
+    flag.Parse()
+}
+```
+
+---
+
+# Secrets 
+
+Add the following code
+
+```go
+func main() {
+    var secret string
+    flag.StringVar(&secret, "secret", "", "DarkSky API key (required)")
+    port := flag.String("port", "3001", "Port service will listen on")
+    flag.Parse()
+
+    // if you want to use Gin
+    if len(secret) == 0 {
+        secret = os.Getenv("secret")
+    }
+    if len(secret) == 0 {
+        flag.PrintDefaults()
+        os.Exit(1)
+    }
+    fmt.Printf("Listening on :%s\n", *port)
+}
+```
+
+---
+
+### Test it
+```bash
+go run . # should fail
+```
+
+```bash
+go run . -secret=foo 
+```
+
+---
+
+# Web API Basics
+
+```go
+package main
+
+import (
+    "net/http"
+)
+
+func main() {
+    http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+        // handle request
+    })
+    http.Handle("/foo", someHandler)  
+    if err := http.ListenAndServe(":3000", nil); err != nil {
+        fmt.Printf("failed to start server: %v", err)
+        os.Exit(1)
+    }
+}
+```
+
+---
+
+# Handlers
+
+## What is a Handler?
+* It's a Go interface that requires the type to have a `ServeHTTP(w, r)` method
+
+## Why use it?
+* Makes it easier to inject dependencies
+
+---
+
+# Listen and Serve 
+
+```go
+type locationHandler struct { }
+
+func (lh locationHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+    fmt.Fprintf(w, "Hello World")
+}
+
+func main() {
+    //...
+    locationHandler := locationHandler{}
+    http.Handle("/", locationHandler)
+    if err := http.ListenAndServe(":"+*port, nil); err != nil {
+        fmt.Printf("failed to start server: %v", err)
+        os.Exit(1)
+    }
+}
+```
+
+```bash
+go run . -secret=abc123 # secret=abc123 gin .
+```
+
+---
+
+# Weather (aka Location) Handler
+
+http://localhost:3000?lat=53.5&lng=-113.3
+
+---
+
+# Querystring params
+
+```go
+q := r.URL.Query()
+lat, err := strconv.ParseFloat(q.Get("lat"), 10)
+if err != nil {
+    http.Error(w, "Invalid lat querystring param", http.StatusBadRequest)
+    return
+}
+```
+
+## Exercise
+
+Parse out the longitude value as well, then test it via the browser
```

Comment:
Maybe add a gist for the code on the slide, or do you want people to type it all out?

Reply:
I really should've had gists for this presentation; next time :)